pyro/nn/module.py

```diff
@@ -826,3 +827,30 @@ def __set__(self, obj, value):
 
 
 PyroModule[torch.nn.RNNBase]._flat_weights = _FlatWeightsDescriptor()
+
+
+# pyro module list
+# using pyro.nn.PyroModule[torch.nn.ModuleList] can cause issues when
+# slice-indexing nested PyroModuleLists, so we define a separate PyroModuleList
+# class that overwrites the __getitem__ method to return a torch.nn.ModuleList
+# to not use self.__class__ in __getitem__, as that would call the
+# PyroModule.__init__ without the parent module context, leading to a loss
+# of the parent module's _pyro_name, and eventually, errors during sampling
+# as parameter names may not be unique anymore
+# The scenario is rare but happend.
+# The fix could not be applied in torch directly, which is why we have to deal
+# with it here, see https://github.com/pytorch/pytorch/issues/121008
+class PyroModuleList(PyroModule, torch.nn.ModuleList):
```

Comment:
This looks great. I think we can add a line to register your new module
```py
_PyroModuleMeta._pyro_mixin_cache[torch.nn.ModuleList] = PyroModuleList
```
That way users can simply type `PyroModule[torch.nn.ModuleList]`, and your fix will be applied to all existing uses of `PyroModule[torch.nn.ModuleList]` including when use in components of existing modules via `to_pyro_module_(...)`.

Reply:
I adressed that in fae8ba6.
I added the first recommendation at the end of the Module.py file, it that was the intention.
I also added the recommended assertion in the tests. The question is whether we keep the tests, which I used to demonstrate the behavior, or if we remove the part that served as a failing example.

Reply:
Updated it via 7516493.
I added the
```python
_PyroModuleMeta._pyro_mixin_cache[torch.nn.ModuleList] = PyroModuleList
```
at the end of the module.py file, if that was the intention.

I also added the assertion to the tests. However, I guess we can remove the "failing test" from the PR (as with the above change, it does not fail anymore). Or should we keep it as a demonstration of usage?